spec/abi.dd

```diff
@@ -957,32 +957,70 @@ $(H2 $(LNAME2 garbage_collection, Garbage Collection))
 
         $(P The interface to this is found in Druntime's $(DRUNTIMESRC core/gc/gcinterface.d).)
 
-$(H2 $(LNAME2 runtime_helper_functions, Runtime Helper Functions))
+$(H2 $(LNAME2 ModuleInfo, ModuleInfo Instance))
 
-        $(P These are found in Druntime's $(DRUNTIMESRC rt/).)
+        $(P An instance of $(LINK2 https://dlang.org/phobos/object.html#.ModuleInfo, `ModuleInfo`)
+        is generated by the compiler and inserted into the object file for every module.
+        `ModuleInfo` contains information about the module that is useful to the D runtime library:
+        )
+
+        $(UL
+        $(LI If the module has a static constructor, static destructor, shared static constructor, or shared static destructor.)
+        $(LI A reference to any unit tests defined by the module.)
+        $(LI An array of references to any imported modules that have one or more of:
+            $(OL
+            $(LI static constructors)
+            $(LI static destructors)
+            $(LI shared static constructors)
+            $(LI shared static destructors)
+            $(LI unit tests)
+            $(LI transitive imports of any module that contains one or more of 1..5)
+            $(LI order independent constructors (currently needed for implementing
+            $(DDSUBLINK dmd, switch-cov, $(TT -cov))))
+            )
+            This enables the runtime to run the unit tests,
+            the module constructors in a depth-first order,
+            and the module destructors in the reverse order.
+        )
+
+        $(LI An array of references to `ClassInfo` for each class defined in the module.
+        $(NOTE this feature may be removed.))
+        )
+
+        $(P `ModuleInfo` is defined in Druntime's $(DRUNTIMESRC object.d).)
```

Comment:
I missed this before, but this statement is also misleading. There is a definition there, but it must match what the compiler generates, since the compiler doesn't consult the runtime lib for the layout. It will suppress generation entirely if ModuleInfo isn't defined there, but if it is, you *must* match the compiler's output in druntime, not vice versa. Meaning the canonical definition is actually in the compiler (and otherwise unspecified right now).

Reply:
> but it must match what the compiler generates

Of course. All compiler generated data follows an external specification. I don't know how to have the compiler read the D spec and follow it, either, but the spec defines the language.

> It will suppress generation entirely if ModuleInfo isn't defined there

The compiler gives a fatal error for that.

Reply:
> The compiler gives a fatal error for that.

No, it doesn't.

Reply:
testmi.d
```
int main() { return 0; }
static this() {}
```

object.d
```
module object;

struct ModuleInfo {} // the definition here is empty, so what do we get?

extern(C) int main() { return 0; }
```

$ dmd testmi.d object.d -c -defaultlib=
// note that there is no error
$ objdump -x testmi.o | grep ModuleInfo
0000000000000000 g     O .data.rel.ro   0000000000000017 _D6testmi12__ModuleInfoZ
0000000000000000 R_X86_64_64       _D6testmi12__ModuleInfoZ

And the ModuleInfo is generated with a non-zero size.


Now comment out that definition of object.ModuleInfo. recompile and check the objdump again, find no ModuleInfo was generated now.

This is predictable from the source code:

glue.d ~ line 600
```
     /* Generate module info for templates and -cov.
     *  Don't generate ModuleInfo if `object.ModuleInfo` is not declared or
     *  explicitly disabled through compiler switches such as `-betterC`.
     *  Don't generate ModuleInfo for C files.
     */
    if (global.params.useModuleInfo && Module.moduleinfo && !m.isCFile /*|| needModuleInfo()*/)
        genModuleInfo(m);
```


There's no check that the ModuleInfo layout matches, there's no querying of the runtime library for the layout. No error generation. It just gets suppressed if it isn't defined, and output in a static format if it is defined.

(btw i have no real problem with this behavior, i actually think it is fine. but the spec shouldn't propagate false information)

Reply:
I meant if ModuleInfo wasn't defined in object.d.

Reply:
Sorry about the confusion.